File: examples/cpp/Beta/tof_neural_fusion_pointcloud.cpp
#include <csignal>
#include <iostream>
#include <opencv2/imgproc.hpp>
#include <optional>

#include "depthai/depthai.hpp"
#include "depthai/remote_connection/RemoteConnection.hpp"

namespace {
volatile std::sig_atomic_t isRunning{1};
void signalHandler(int) {
    isRunning = 0;
}

class DepthColorizer : public dai::NodeCRTP<dai::node::HostNode, DepthColorizer> {
   public:
    Input& input = inputs["depth"];

    std::shared_ptr<DepthColorizer> build(Output& depth) {
        depth.link(input);
        return std::static_pointer_cast<DepthColorizer>(this->shared_from_this());
    }

    std::shared_ptr<dai::Buffer> processGroup(std::shared_ptr<dai::MessageGroup> messages) override {
        auto colorized = dai::utility::colorizeDepthFrame(*messages->get<dai::ImgFrame>("depth"));
        auto color = colorized.getCvFrame();
        cv::cvtColor(color, color, cv::COLOR_BGR2RGB);
        colorized.setCvFrame(color, dai::ImgFrame::Type::RGB888i);
        return std::make_shared<dai::ImgFrame>(std::move(colorized));
    }
};
}  // namespace

int main() {
    std::signal(SIGINT, signalHandler);

    dai::RemoteConnection remote;
    dai::Pipeline pipeline;
    auto left = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B, std::nullopt, 30.0f);
    auto right = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C, std::nullopt, 30.0f);
    auto fusion = pipeline.create<dai::beta::node::ToFStereoFusion>()->build(left, right);

    auto colorizer = pipeline.create<DepthColorizer>()->build(fusion->depth);
    auto pointcloud = pipeline.create<dai::node::PointCloud>();
    pointcloud->setRunOnHost(true);
    fusion->depth.link(pointcloud->inputDepth);
    colorizer->out.link(pointcloud->getColorInput());
    remote.addTopic("pcl", pointcloud->outputPointCloud);

    pipeline.start();
    remote.registerPipeline(pipeline);
    std::cout << "Visualizer running at http://localhost:8082\n";

    while(isRunning != 0 && pipeline.isRunning() && remote.waitKey(1) != 'q') {
    }
}

File: examples/python/Beta/tof_neural_fusion_pointcloud.py
#!/usr/bin/env python3
"""Fused ToF and neural-depth point cloud in the DepthAI Visualizer.

Open http://localhost:8082 after starting this script.
"""

import cv2
import depthai as dai


class DepthColorizer(dai.node.ThreadedHostNode):
    def __init__(self):
        super().__init__()
        self.input = self.createInput()
        self.output = self.createOutput()

    def run(self):
        while self.isRunning():
            colorized = dai.utility.colorizeDepthFrame(self.input.get())
            colorized.setCvFrame(cv2.cvtColor(colorized.getCvFrame(), cv2.COLOR_BGR2RGB), dai.ImgFrame.Type.RGB888i)
            self.output.send(colorized)


with dai.Pipeline() as pipeline:
    left = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B, sensorFps=30)
    right = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C, sensorFps=30)
    fusion = pipeline.create(dai.beta.node.ToFStereoFusion).build(left, right)
    fusion.initialConfig.confidenceThreshold = 0.0
    fusion.neuralDepth.initialConfig.setConfidenceThreshold(0)

    colorizer = pipeline.create(DepthColorizer)
    fusion.depth.link(colorizer.input)

    pointcloud = pipeline.create(dai.node.PointCloud)
    pointcloud.setRunOnHost(True)
    fusion.depth.link(pointcloud.inputDepth)
    colorizer.output.link(pointcloud.inputColor)

    remote = dai.RemoteConnection()
    remote.addTopic("pcl", pointcloud.outputPointCloud, "3d")

    fusion_config_queue = fusion.inputConfig.createInputQueue()
    neural_config_queue = fusion.neuralDepth.inputConfig.createInputQueue()
    tof_config_queue = fusion.tof.tofBaseNode.inputConfig.createInputQueue()

    def set_fusion_confidence(value):
        fusion.initialConfig.confidenceThreshold = value / 100.0
        fusion_config_queue.send(fusion.initialConfig)

    def set_neural_confidence(value):
        fusion.neuralDepth.initialConfig.setConfidenceThreshold(value)
        neural_config_queue.send(fusion.neuralDepth.initialConfig)

    def set_temporal_filter(value):
        fusion.tof.tofBaseNode.initialConfig.enablePhaseShuffleTemporalFilter = bool(value)
        tof_config_queue.send(fusion.tof.tofBaseNode.initialConfig)

    pipeline.start()
    cv2.namedWindow("Confidence thresholds", cv2.WINDOW_NORMAL)
    cv2.createTrackbar("Fusion confidence (%)", "Confidence thresholds", 0, 100, set_fusion_confidence)
    cv2.createTrackbar("Neural confidence", "Confidence thresholds", 0, 255, set_neural_confidence)
    cv2.createTrackbar("Temporal filter", "Confidence thresholds",
                       int(fusion.tof.tofBaseNode.initialConfig.enablePhaseShuffleTemporalFilter), 1, set_temporal_filter)
    remote.registerPipeline(pipeline)
    print("Visualizer running at http://localhost:8082")

    while pipeline.isRunning():
        if cv2.waitKey(1) == ord("q") or remote.waitKey(1) == ord("q"):
            break

cv2.destroyAllWindows()
